testing/scenario_app/bin/utils/process_manager_extension.dart

```diff
@@ -8,35 +8,51 @@ import 'dart:io';
 
 import 'package:process/process.dart';
 
-/// Pipes the [process] streams and writes them to [out] sink.
-/// If [out] is null, then the current [Process.stdout] is used as the sink.
-Future<int> pipeProcessStreams(
-  Process process, {
-  StringSink? out,
-}) async {
-  out ??= stdout;
+(Future<int> exitCode, Stream<String> output) getProcessStreams(
+  Process process,
+) {
   final Completer<void> stdoutCompleter = Completer<void>();
-  final StreamSubscription<String> stdoutSub = process.stdout
+  final Completer<void> stderrCompleter = Completer<void>();
+  final StreamController<String> outputController = StreamController<String>();
+
+  final StreamSubscription<void> stdoutSub = process.stdout
     .transform(utf8.decoder)
     .transform<String>(const LineSplitter())
-    .listen((String line) {
-      out!.writeln('[stdout] $line');
-    }, onDone: stdoutCompleter.complete);
+    .listen(outputController.add, onDone: stdoutCompleter.complete);
 
-  final Completer<void> stderrCompleter = Completer<void>();
-  final StreamSubscription<String> stderrSub = process.stderr
+  // From looking at historic logs, it seems that the stderr output is rare.
+  // Instead of prefacing every line with [stdout] unnecessarily, we'll just
+  // use [stderr] to indicate that it's from the stderr stream.
+  //
+  // For example, a historic log which has 0 occurrences of stderr:
+  // https://gist.github.com/matanlurey/84cf9c903ef6d507dcb63d4c303ca45f
+  final StreamSubscription<void> stderrSub = process.stderr
     .transform(utf8.decoder)
     .transform<String>(const LineSplitter())
-    .listen((String line) {
-      out!.writeln('[stderr] $line');
-    }, onDone: stderrCompleter.complete);
+    .map((String line) => '[stderr] $line')
+    .listen(outputController.add, onDone: stderrCompleter.complete);
 
-  final int exitCode = await process.exitCode;
-  await stderrSub.cancel();
-  await stdoutSub.cancel();
+  final Future<int> exitCode = process.exitCode.then<int>((int code) {
+    stdoutSub.cancel();
+    stderrSub.cancel();
+    outputController.close();
+    return code;
```

Comment:
oh wait, should we not await the cancel futures before closing the outputController?

Reply:
![giphy (23)](https://github.com/flutter/engine/assets/30870216/7bb96061-9469-4eb5-996e-9eff65803f97)

Reply:
my vague understanding of how this works is that stdout/stderr data can come in AFTER the exit code.

Reply:
actually now, I wonder if we should be awaiting for these subs to complete first. just cancelling them might lead to dropped logs. See some prior art here: https://github.com/flutter/flutter/blob/master/packages/flutter_tools/lib/src/base/process.dart#L539

Reply:
Done!